### ISSUE TITLE
Broken Search Box on Discover page

### ISSUE THREAD

**Issue by Lizabrazil** [hydroshare/hydroshare#2308]
When I type "industrial" into the text search box on https://www.hydroshare.org/search/ I should have this resource returned: https://www.hydroshare.org/resource/4800f6fafe7e43bc969a3770ea817738/

Only when I narrow resources down by owner or author does this resource appear. 
Seems the text search functionality is broken. 
 
![image](https://user-images.githubusercontent.com/13852599/30132751-808e5dac-931f-11e7-818a-e0be0393c118.png)

**Comment by alvacouch** [comment]
@Lizabrazil @dtarb @martinseul @mjstealey @cbcunc

My initial theory is that the nightly updates to the SOLR catalog aren't happening. This should show up if they were up to date. 

* My first suggestion is to first run the manual SOLR catalog building procedure that Michael used to build the catalog from scratch, and see if that addresses this particular problem. 
* If that fixes the problem, I have the ability to reproduce it by creating a new resource. 
* If that does not fix the problem, then there is a schema problem. This is very, very unlikely.

**Comment by alvacouch** [comment]
@dtarb @martinseul @Lizabrazil @mjstealey @cbunc The manual catalog update process does not seem to be running now. We should turn that back on. However, the bug seems to have two parts: 

1. The overnight update is needed and isn't running. 
2. The immediate post of the resource to SOLR when it's made public does not include enough metadata for basic discovery. 

A reminder of how this is supposed to work. 

1. When a resource is made public, a half-hearted SOLR record is generated and pushed to SOLR. It is not necessarily a full record. Mostly we did this to save development time. It's enough to discover the resource but not to discover it in all ways. 
2. Theoretically, overnight, these records are updated to be full records. 
3. When a resource is made private, its SOLR record is immediately deleted. 

So the nightly update needs to be turned back on, and the temporary push to SOLR needs to be made a bit more complete so that basic textbox searching works immediately.

**Comment by alvacouch** [comment]
@phuongdm This is where we're discussing the SOLR problems.

**Comment by cbcunc** [comment]
I have enabled Jenkins->Housekeeping->nightly-rebuild-solr-index. It's set to run between midnight and 1am nightly. But I just manually, initiated it. I don't know how long it should take to run. But it seemed to be sitting and spinning, as well as running a Django shell, so I killed the job and investigated. hsctl It appears to only be running `docker exec hydroshare python manage.py rebuild_index --noinput` in response to a reubuild_index subcommand. This doesn't seem quite right to me. I would have thought it would be talking to the solr container.

**Comment by cbcunc** [comment]
There is no indication this job has ever run before through Jenkins. How have the index rebuilds been occurring all these years?

**Comment by alvacouch** [comment]
@cbcunc I have no idea! 

This may have been dropped from Jenkins for some reason.

**Comment by alvacouch** [comment]
@dtarb @Lizabrazil @martinseul @aphelionz I analyzed the code and it is really easy to duplicate the behavior that @Lizabrazil describes:

1. Create a resource with the wrong name. 
2. Make it public. 
3. Change the title to have a desired keyword that doesn't appear anywhere else. 

That keyword will not be discoverable using the subject box. The reason is that SOLR data isn't updated except when something is made public or private. For now, a simple fix is to make the resource private and then make it public again. This will push the change to SOLR. Longer term, we need to insure that SOLR is updated for any change of a public resource. 

There are other deeper issues: SOLR is updated when a user explicitly changes something to private, but is not updated when a resource is forced to be private due to inappropriate edits. Solution to this is to write a definitive routine for making a resource private that always updates SOLR, and then call it everywhere one wants something to become private.

**Comment by dtarb** [comment]
Thanks for tracking this down.

**Comment by phuongdm** [comment]
@alvacouch:

From 10:00am to 3:00pm this Wednesday, I and @yxiao did some investigations on the "SOLR unsecurity link" and found that "Last Modified" field of "Collection1" Core were changed several times. And, I agree with @cbcunc that the Jenkins's task didn't run on that time (in fact it never run before).

Is that the case you wrote: "SOLR is updated when a user explicitly changes something to private"?

**Comment by phuongdm** [comment]
@alvacouch , @cbcunc 

+ ./hsctl managepy rebuild_index ===> asked me to make sure delete all docs or not ===> we can't call it in Jenkins task

+ ./hsctl rebuild_index  ===> run without asking  ===> could call in Jenkins task

On cuahsi-dev-1.hydroshare.org, it took me around 12 minutes to finish. I will manually run it tonight on www.hydroshare.org and see how long will it take then decide to run the task daily or weekly.

**Comment by cbcunc** [comment]
./hsctl rebuild_index is the correct command to run (no managepy). Running it does not fix the problem in this issue though. I spoke with @mjstealey and rebuild_index has only ever been run by ./hsctl rebuild when rebuilding the site. It doesn't appear the intention was to run it every day or on a regular schedule, but maybe on demand from Jenkins.

**Comment by phuongdm** [comment]
Make sense, thanks

**Comment by Lizabrazil** [comment]
I still cannot find the resource...
I need this to work next week for a class exercise. I typed "industrial" @alvacouch 
Thanks

**Comment by mjstealey** [comment]
@Lizabrazil - Oddly typing in a portion of the word, like "indus" seems to find it.

![screen shot 2017-09-27 at 3 40 10 pm](https://user-images.githubusercontent.com/5332509/30934032-45aaea6c-a39a-11e7-9a32-51e88faa7e0f.png)

**Comment by dtarb** [comment]
@Lizabrazil for the workshop can you fix it by making it private then public as @alvacouch suggested.

**Comment by Lizabrazil** [comment]
@dtarb That did not work. I need this fixed by October 6th or I'll have to take out the HydroShare portion of the exercise.

**Comment by Lizabrazil** [comment]
@alvacouch If i search just for "plant" then i can find the resource.

**Comment by alvacouch** [comment]
@lizabrazil That suggests that the search query is not matching the first word in the title. Could you do me a big favor and see if that happens for other resources?

**Comment by alvacouch** [comment]
This line is very, very suspicious 

[line 19 of hs_core/discovery_form.py](https://github.com/hydroshare/hydroshare/blob/develop/hs_core/discovery_form.py#L19)

This is the line: 

```
sqs = self.searchqueryset.filter(text__startswith=self.cleaned_data.get('q'))\
               .filter(discoverable=True)\
               .filter(is_replaced_by=False)
```

The line includes the query from the search box (named 'q'). The query asserts that the text field "starts with" the query q. Since the text field is a catchall field, it contains all text from all fields in a non-deterministic order. Thus `startswith` is most definitely the wrong operator.

Looking deeper, the text field is: 
"text": " \n    4800f6fafe7e43bc969a3770ea817738\n \n \n\n    Industrial Plants PA\n \n \n    This layer  displays the coordinates of several industrial waste plants throughout Pennsylvania.\n \n\n\n\nMay 4, 2017, 5:36 p.m.\n\n \n \n \n    eng\n \n\n    GenericResource\n \n \n    Generic\n \n \n \n \n \n\n    \n        liza brazil\n     \n\n\n\n    \n        pennsylvania\n     \n\n    \n        industrial\n     \n\n    \n        industrial plants\n     \n\n    \n        water quality\n     \n\n\n    \n        CUAHSI\n     \n\n\n    \n        lbrazil@cuahsi.org\n    \n\n\n\n\n\n\n\n\n\n\n    application/json\n\n\n    hydroShareIdentifier\n\n\n\n\n    lizabrazil\n\n\n    liza brazil\n\n\n    liza brazil\n\n\n    \n\n\n    lizabrazil\n\n\n    liza brazil\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n",
        
This is really ugly and could have been cleaned up by deleting \ns but it is clear that **"Industrial Plants PA" is not the first string!**


Thus this would result in the behavior we see.

**Comment by alvacouch** [comment]
Several things here are suspicious. 

The code `startswith` would be fine if the text field were tokenized. However, it is not tokenized; it is a generic character field: 

[Line 16 of search_indexes.py](https://github.com/hydroshare/hydroshare/blob/develop/hs_core/search_indexes.py#L16)

Fixing that requires changing the declarations and re-indexing. 

To fix this temporarily without re-indexing, see the following page: 
http://django-haystack.readthedocs.io/en/master/searchqueryset_api.html#field-lookups
We need to replace `startswith` with something more reasonable. 

Let's weigh the options: 
* `contains` will match substrings of words, e.g., "logical" matches "hydrological". Users didn't like that before, which is how we got this solution. 
* `fuzzy` works like `contains` but matches near misspellings. Even more strange results. 

For now, a simple solution is to use `contains'. A better solution would be to switch text to tokenized and then use `startswith`, which will do what we really want. 

My guess is that we got to this particular solution by not reading a stackoverflow post carefully enough.

**Comment by alvacouch** [comment]
@dtarb @Lizabrazil @martinseul The above proposal is not going to fix this. Somehow, the keywords are being tokenized and stemmed internally. I tried other words and the situation was consistent. For example, "variat" matches "variations" but "variation" does not. Likewise, "indus" matches "industrial" but "industria" does not. So, the search is stemming the words that come into text and then matching the stems. 

"indus plant" works but "indus plants" does not. "plant" is the stem of "plants". 

Solution is to turn stemming off.

**Comment by alvacouch** [comment]
@dtarb @Lizabrazil @martinseul 
Here is the root of the problem: 

```
    <fieldType name="text_en" class="solr.TextField" positionIncrementGap="100">
      <analyzer type="index">
        <tokenizer class="solr.StandardTokenizerFactory"/>
        <filter class="solr.StopFilterFactory"
                ignoreCase="true"
                words="lang/stopwords_en.txt"
                enablePositionIncrements="true"
                />
        <filter class="solr.LowerCaseFilterFactory"/>
        <filter class="solr.EnglishPossessiveFilterFactory"/>
        <filter class="solr.KeywordMarkerFilterFactory" protected="protwords.txt"/>
        <!-- Optionally you may want to use this less aggressive stemmer instead of PorterStemFilterFactory:
          <filter class="solr.EnglishMinimalStemFilterFactory"/>
        -->
        <filter class="solr.PorterStemFilterFactory"/>
      </analyzer>
```
This is applied to the text field (on beta). This reads that all words appearing in the whole resource are transformed by removing possessives, reducing to the stem of the word, etc.  The problem is that when you make a query, including one of these unmodified words doesn't match.

**Comment by alvacouch** [comment]
@dtarb @martinseul @Lizabrazil There is still something strange about this. Theoretically, both the query and the text should be stemmed. This doesn't seem to be happening. It seems that the query is not being so transformed. Perhaps this is the result of "startswith"; I don't know. 

Just so you understand a bit more, here is some background. 

"Stemming" is the process of deriving a root word from a complex word. Some stemmers are simple, and some are quite complex. The simplest transform "papers" to "paper" (removing plurals) and "plastering" to "plaster" (or "plast"). 

The point is that everything works properly if both the query and the text to be matched are stemmed in the same way. Somehow, 

* the text is getting stemmed.
* the query is not getting stemmed.  

This is happening automagically in Haystack. Queries direct to SOLR are working properly. 

So the whole scheme is breaking down. 

The root of the problem may be here: 

[line 19 of hs_core/discovery_form.py](https://github.com/hydroshare/hydroshare/blob/develop/hs_core/discovery_form.py#L19)

In this line, the query "startswith" is used rather than "contains". I think that might have caused this behavior (because stemming a prefix doesn't make sense). 

Could someone try "startswith" => "contains" to see if that addresses the issue?

**Comment by pkdash** [comment]
May be better to use icontains (https://docs.djangoproject.com/en/dev/ref/models/querysets/#icontains)

**Comment by alvacouch** [comment]
@pkdash icontains is not implemented for haystack. Haystack is an incomplete implementation of the queryset model for Django.

**Comment by alvacouch** [comment]
@martinseul @dtarb @Lizabrazil @pkdash 

My research suggests that the simplest fix is to replace `text__startswith=` with `text=`. This will invoke the default mechanism for matching in SOLR, which is term-by-term aggressive stem matching.  

What this will do: Everything will be tokenized and matched word-to-word (both queries and titles): 

* "Industrial" matches "Industrial"
* "Indus" doesn't match "Industrial" 
* "Indust" does match "Industrial" 

because stem("Industrial") == "Indust". 

What we really want is "text= or text__startswith=". Django Q expressions do not work for Haystack. Instead, Haystack supports "SQ" expressions. The appropriate expression might be 
```
cleaned_q = self.cleaned_data.get('q')
sqs = self.searchqueryset.filter(SQ(text__startswith=cleaned_q) | SQ(text=cleaned_q))\
                .filter(discoverable=True)\
                .filter(is_replaced_by=False)
 ```
What this does: 

| query | stem(query) | stored | stem(stored) | matches? | how? |
|--------|------|------|------|-----|------|
| Industrial | indust | Industrial | indust | Yes | = |
| Indust | indust | Industrial | indust | Yes | = |
| Indus | indus | Industrial | indust | Yes | startswith | 
| Industr | industr | Industrial | indust | No | not prefix or exact |

So, things match unless one specifies a match term that is between the stem length and the total length. Then matches fail.  

Other than this, we are getting beyond what Haystack was designed to do and have to take more drastic measures.

**Comment by alvacouch** [comment]
@Lizabrazil @martinseul @aphelionz @dtarb 
The proposed fix is running on cuahsi-dev-2.hydroshare.org. Here is the current behavior (note the stem correction from the above): 

| query | stem(query) | stored | stem(stored) | matches? | how? |
|--------|------|------|------|-----|------|
| Industrial | industri | Industrial | industri | Yes | = |
| Industria | industria | Industrial | industri | No | neither prefix nor = |
| Indus | indus | Industrial | industri | Yes | prefix of stem but not =| 
| Industri | industri | Industrial | industri | Yes | prefix of stem but not = |

So, "Industrial" and "Industri" match, but "Industria" does not match due to the aggressiveness of the stemming algorithm. 

I will submit a PR.

**Comment by alvacouch** [comment]
@dtarb @Lizabrazil @martinseul @aphelionz 
If this isn't good enough, I suggest using the weak stemmer instead of the aggressive stemmer that is the default in SOLR. This requires patching the SOLR generator in Haystack. This is common enough a task that there are instructions on the web as to how to do it. This then generates a SOLR configuration file that uses a weaker stemmer, that stems "`planned`" => "`plan`" and "`grapes`" => "`grape`" and "`Alva's`" to "`Alva`", but does **not** stem "`industrial`" to "`industri`".

Then all prefixes of "Industrial" will match "Industrial".

**Comment by alvacouch** [comment]
This is fixed in 1.13.1.

**Comment by aphelionz** [comment]
@alv
